### ISSUE TITLE
Wrong error on wrapping arithmetic with -Zmir-opt-level=2

### ISSUE THREAD

**Issue by bytwise** [rust-lang/rust#82646]
I tried this code with `-Z mir-opt-level=2`:

```rust
fn main() {
    println!("{}", 1_u32.wrapping_sub(2_u32));
}
```

It compiles and runs without mir-opt, but with `-Z mir-opt-level=2` or above I get:

```
error: this arithmetic operation will overflow
   --> src/main.rs:2:13
    |
2   |       println!("{}", 1_u32.wrapping_sub(2_u32));
    |                      ^^^^^^^^^^^^^^^^^^^^^^^^^ attempt to compute `1_u32 - 2_u32`, which would overflow
    |
```

### Meta

`rustc --version --verbose`:
```
rustc 1.52.0-nightly (0148b971c 2021-02-18)
binary: rustc
commit-hash: 0148b971c921a0831fbf3357e5936eec724e3566
commit-date: 2021-02-18
host: x86_64-apple-darwin
release: 1.52.0-nightly
LLVM version: 11.0.1
```

**Comment by leonardo-m** [comment]
I don't even understand the asm for just "-O" compilation of that code.

**Comment by osa1** [comment]
Wrapping arithmetic operators are lowered here: https://github.com/rust-lang/rust/blob/cbca5689a5a0c63c6c5fda22bb0678164b52fec3/compiler/rustc_mir/src/transform/lower_intrinsics.rs#L43-L67

Looking at the type of MIR binary operators: https://github.com/rust-lang/rust/blob/cbca5689a5a0c63c6c5fda22bb0678164b52fec3/compiler/rustc_middle/src/mir/mod.rs#L2116-L2152

This type doesn't have distinct operators for wrapping and checked operations, so in `check_binary_op`: https://github.com/rust-lang/rust/blob/cbca5689a5a0c63c6c5fda22bb0678164b52fec3/compiler/rustc_mir/src/transform/const_prop.rs#L569-L627

(which is where we evaluate the binary op and report an overflow)

As far as I can see there is no way to tell whether overflowing is OK or not.

Should we split MIR binary operators into wrapping and overflowing variants?
